Made Armor upgradable

reaperrr opened the pull request:
This allows to change armor on levelup, and even add additional armor types.

reaperrr committed:
Made Armor upgradable

reaperrr commented:
Note: While not strictly a requirement, it would be better if #9006 was merged first so I can get rid of the `CanTargetActor` work-arounds.

Mailaender commented:
You could set up the crates. The animation sequences have already been identified:

* https://github.com/OpenRA/OpenRA/blob/bleed/mods/cnc/sequences/misc.yaml#L168
* https://github.com/OpenRA/OpenRA/blob/bleed/mods/ts/sequences/misc.yaml#L52
* https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/sequences/misc.yaml#L312

reaperrr commented:
> You could set up the crates. The animation sequences have already been identified:

The 'armor' crates in the originals simply halved the damage taken, which is already possible (and easier to do) with the `DamageMultiplier` trait.

I don't mind setting up armor crates, but not as part of this PR.

pchote commented:
FYI we used to have the separate stat crates in RA and TD, but we decided to replace them all with the veterancy crate because individual stat upgrades don't really make sense for us (remember: the original games didn't have any veterancy).

GraionDilach commented:
TS/RA2 had both. The crate-given individual stat was bigger than the stat one would gain from ranking.

reaperrr commented:
Rebased.

GraionDilach commented on OpenRA.Mods.Common/Warheads/DamageWarhead.cs:
```diff
@@ -34,13 +35,13 @@ public static object LoadVersus(MiniYaml yaml)
+			var armor = victim.TraitsImplementing<Armor>().Where(a => !a.IsTraitDisabled && a.Info.Type != null);
+			foreach (var a in armor)
 			{
 				int versus;
```
Move the creation of this out from the foreach, with a default value of 100 and multiply it here to let more armors stack with each other, also move the return out from the foreach too.

reaperrr replied:
I've now wasted 4 hours trying to make this work with no success, so unless someone else comes up with some concrete, working code, there's no telling when this will be finished.

reaperrr commented:
I don't know how to handle IEnumerable and Versus.TryGetValue in order to make armor effects stack. Wasted over 5 hours so far without success.

GraionDilach commented:
Give me tomorrow.

Commit:
Fix armors.

GraionDilach commented:
Filed https://github.com/reaperrr/OpenRA/pull/9.

reaperrr committed:
Update outdated description

reaperrr commented:
> Filed reaperrr#9.

Looks so simple...
Thanks, updated.

Also updated test case (disabled veterancy damage modifier at first two rank-ups, so any changes to received damage are entirely caused by the armor).

reaperrr commented:
Explanation on test case for reviewers:

The correct behavior when shooting at the allied jeep should be

- medium/heavy tank deal 75% of max. damage before rank-up, 100% after both 1st and 2nd rank-up (at 2nd rank-up, the jeep has two 'layers' of heavy armor, but cannons have 100% versus heavy, so 100% * 100% = 100%).

- light tank does 100% damage befor rank-up, 40% after 1st rank-up, 16% after 2nd rank-up (40% * 40%).

GraionDilach commented:
I don't know how my :+1: would be valid after that patch... but yea, I say :+1:

penev92 commented:
This removes the `.TryGetValue()` safety net, but I'm OK with that.
Works :+1:

reaperrr commented:
> I don't know how my :+1: would be valid after that patch... but yea, I say :+1:

Well, I reviewed and tested your patch, and I assume you tested it as well, so in my opinion this PR has +2 already.

penev92 commented:
Good enough for me. I'll merge when you remove the test commits.

reaperrr commented:
Test case removed.

penev92 commented:
Thanks.
